### PR TITLE
Pass env to ComposeDelegate in DockerComposeContainer#stop

### LINKED ISSUE #8492: [Bug]: DockerComposeContainer: Stop exception when variable substitution is used in compose file
### Module

Core

### Testcontainers version

1.19.7

### Using the latest Testcontainers version?

Yes

### Host OS

Linux Debian/sid

### Host Arch

amd64

### Docker version

```shell
Client: Docker Engine - Community
 Version:           26.0.0
 API version:       1.45
 Go version:        go1.21.8
 Git commit:        2ae903e
 Built:             Wed Mar 20 15:18:01 2024
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          26.0.0
  API version:      1.45 (minimum version 1.24)
  Go version:       go1.21.8
  Git commit:       8b79278
  Built:            Wed Mar 20 15:18:01 2024
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.28
  GitCommit:        ae07eda36dd25f8a1b98dfbf587313b99c0190bb
 runc:
  Version:          1.1.12
  GitCommit:        v1.1.12-0-g51d5e94
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
 rootlesskit:
  Version:          2.0.2
  ApiVersion:       1.1.1
  NetworkDriver:    slirp4netns
  PortDriver:       builtin
  StateDir:         /run/user/1000/dockerd-rootless
 slirp4netns:
  Version:          1.2.1
  GitCommit:        09e31e92fa3d2a1d3ca261adaeb012c8d75a8194
```

### What happened?

I have a docker compose file which contains a variable substitution (${VAR}). A DockerComposeContainer is created to use the compose file and set the variable a withEnv-call. This works fine, but as soon the stop is called, an Exception is thrown due to an error (IllegalStateException: Container did not start correctly.).

```yaml
version: "3.9"
services:
  mock:
    image: wiremock/wiremock:3.0.3-1
    hostname: mock
  db:
    image: postgres:16
    hostname: db
networks:
  default:
    external:
      name: ${NETWORK}
```

```java
Network containerNetwork = Network.newNetwork()
DockerComposeContainer composeContainer = new DockerComposeContainer("compose-file.yaml")
            .withExposedService("mock", 8080)
            .withExposedService("db", 5432)
            .withEnv("NETWORK", containerNetwork.getId());
composeContainer.start();
// .... everything fine until now
// the following stop will raise the excpeption, because inside of DockerComposeContainer#stop the env is not passed to the composeDelegate
composeContainer.stop();

```

### Relevant log output

```shell
Caused by: org.testcontainers.containers.ContainerLaunchException: Container startup failed for image docker/compose:1.29.2
        at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:359)
        at org.testcontainers.containers.GenericContainer.start(GenericContainer.java:330)
        at org.testcontainers.containers.ContainerisedDockerCompose.invoke(ContainerisedDockerCompose.java:64)
        at org.testcontainers.containers.ComposeDelegate.runWithCompose(ComposeDelegate.java:254)
        at org.testcontainers.containers.ComposeDelegate.runWithCompose(ComposeDelegate.java:240)
        at org.testcontainers.containers.DockerComposeContainer.stop(DockerComposeContainer.java:175)
        at my.custom.DockerComposeResource.stop(DockerComposeResource.java:142)
        at io.quarkus.test.common.TestResourceManager.close(TestResourceManager.java:177)
        ... 49 more
Caused by: org.rnorth.ducttape.RetryCountExceededException: Retry limit hit with exception
        at org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess(Unreliables.java:88)
        at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:344)
        ... 56 more
Caused by: org.testcontainers.containers.ContainerLaunchException: Could not create/start container
        at org.testcontainers.containers.GenericContainer.tryStart(GenericContainer.java:563)
        at org.testcontainers.containers.GenericContainer.lambda$doStart$0(GenericContainer.java:354)
        at org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess(Unreliables.java:81)
        ... 57 more
Caused by: java.lang.IllegalStateException: Container did not start correctly.
        at org.testcontainers.containers.GenericContainer.tryStart(GenericContainer.java:495)
        ... 59 more
```

### Additional Information

I have a fix prepared and will open a PR.

### PR DESCRIPTION
**Pull request by mmusenbr**
This fixes issues on stopping when ENV-Variables are used inside the compose file. Issue was introduced in 7112db5.

Fixes #8492

### CODE REVIEW

**Comment by eddumelendez** [comment]
Thanks for your contribution, @mmusenbr! I wonder why [this test](https://github.com/testcontainers/testcontainers-java/blob/7112db5c6e493306360818d6bfc261df92c49255/core/src/test/java/org/testcontainers/junit/ComposePassthroughTest.java) has not failed. I see during the refactor this was omitted but would be nice to have a test to make sure this is not happening again.

**Comment by mmusenbr** [comment]
Hi @eddumelendez , I need to answer why [ComposePassthroughTest](https://github.com/testcontainers/testcontainers-java/blob/7112db5c6e493306360818d6bfc261df92c49255/core/src/test/java/org/testcontainers/junit/ComposePassthroughTest.java) is passing with two aspects:
- Running via [ComposeContainer](https://github.com/testcontainers/testcontainers-java/blob/7112db5c6e493306360818d6bfc261df92c49255/core/src/main/java/org/testcontainers/containers/ComposeContainer.java) does not seem to need all the details set, as is uses the information from the `COMPOSE_PROJECT_NAME`

For example having the following yaml:
```yaml
services:
  redis:
    image: redis
    ports:
      - ${REDIS_PORT}
```
And then running:
`COMPOSE_PROJECT_NAME=randprojname REDIS_PORT=8272 docker compose -f compose.yml up`
and then:
`COMPOSE_PROJECT_NAME=randprojname docker compose -f compose.yml down`
works just with a warning: `WARN[0000] The "REDIS_PORT" variable is not set. Defaulting to a blank string.`

Whereas running both commands without `COMPOSE_PROJECT_NAME`, the down command fails with `1 error(s) decoding: * 'services[redis].ports[0]' expected a map, got 'string'`.

So, even with this yaml [DockerComposePassthroughTest](https://github.com/testcontainers/testcontainers-java/blob/7112db5c6e493306360818d6bfc261df92c49255/core/src/test/java/org/testcontainers/junit/DockerComposePassthroughTest.java) would pass, BUT not [ComposePassthroughTest](https://github.com/testcontainers/testcontainers-java/blob/7112db5c6e493306360818d6bfc261df92c49255/core/src/test/java/org/testcontainers/junit/ComposePassthroughTest.java)

- Additionally it seems to depend what is substituted to trigger a failure. The used definition in the tests  [DockerComposePassthroughTest](https://github.com/testcontainers/testcontainers-java/blob/7112db5c6e493306360818d6bfc261df92c49255/core/src/test/java/org/testcontainers/junit/DockerComposePassthroughTest.java) and [ComposePassthroughTest](https://github.com/testcontainers/testcontainers-java/blob/7112db5c6e493306360818d6bfc261df92c49255/core/src/test/java/org/testcontainers/junit/ComposePassthroughTest.java) is [v2-compose-test-passthrough.yml](https://github.com/testcontainers/testcontainers-java/blob/7112db5c6e493306360818d6bfc261df92c49255/core/src/test/resources/v2-compose-test-passthrough.yml) which 'only' substitutes an ENV variable. If this is not substituted, it is not enough to bring docker compose to report an error.

For example having the following yaml:
```yaml
services:
  redis:
    image: redis
    environment:
      test=${REDIS_PORT}
```

And running WITHOUT `COMPOSE_PROJECT_NAME` the commands:
`REDIS_PORT=8272 docker compose -f compose.yml up`
and then:
`docker compose -f compose.yml down`
it succeeds and only a warning is printed again.

**Comment by mmusenbr** [comment]
I have added two tests, as described above, the `ComposeContainerPortViaEnvTest` would also not fail with the current code, as only the warning is visible in the logs, but stopping works fine. But `DockerComposeContainerPortViaEnvTest` is the one, which would fail now, but is fixed with this PR.

File: core/src/main/java/org/testcontainers/containers/ComposeContainer.java [hunk 1]
```diff
@@ -165,7 +165,7 @@ public void stop() {
                 if (removeImages != null) {
                     cmd += " --rmi " + removeImages.dockerRemoveImagesType();
                 }
-                this.composeDelegate.runWithCompose(this.localCompose, cmd);
+                this.composeDelegate.runWithCompose(this.localCompose, cmd, env);
```
Comment (eddumelendez) [inline core/src/main/java/org/testcontainers/containers/ComposeContainer.java:168]:
```suggestion
                this.composeDelegate.runWithCompose(this.localCompose, cmd, this.env);
```

File: core/src/main/java/org/testcontainers/containers/DockerComposeContainer.java [hunk 2]
```diff
@@ -172,7 +172,7 @@ public void stop() {
                 if (removeImages != null) {
                     cmd += " --rmi " + removeImages.dockerRemoveImagesType();
                 }
-                this.composeDelegate.runWithCompose(this.localCompose, cmd);
+                this.composeDelegate.runWithCompose(this.localCompose, cmd, env);
```
Comment (eddumelendez) [inline core/src/main/java/org/testcontainers/containers/DockerComposeContainer.java:175]:
```suggestion
                this.composeDelegate.runWithCompose(this.localCompose, cmd, this.env);
```

**Review by eddumelendez — DISMISSED** [review]

**Comment by eddumelendez** [comment]
Thanks for your contribution and detailed explanation, @mmusenbr !

**Pull request merged**